### ISSUE TITLE
[question] Local video correlation (with temporal context length > 1) 

### ISSUE THREAD

**Issue by vadimkantorov** [getkeops/keops#148]
I'm looking for a memory-efficient local spatio-temporal correlation kernel to build an attention-like label propagation for video (similar to https://github.com/ajabri/videowalk label prop).

Is KeOps suitable for this task?

Existing correlation kernels https://github.com/princeton-vl/RAFT/blob/master/alt_cuda_corr/correlation_kernel.cu would require launching a kernel `context_len` times in total to account for all temporal gaps. This is okay, but I am wondering whether KeOps already has such kernels and can solve the task in a neat way (and maybe even fuse with topk and softmax that are subsequently used for label prop)

Thanks @jeanfeydy!

**Comment by jeanfeydy** [comment]
Hi @vadimkantorov ,

Thanks for your very interesting question!
Even if no tutorial on the subject is available yet on our website, KeOps can accelerate two types of operations on grid ("bitmap") images:

- **Separable** operations, such as [distance transforms](https://en.wikipedia.org/wiki/Distance_transform) that can be used to implement fast Chamfer distances or discrete optimal transport solvers.
- **Patch-based** computations, where local neighborhoods interact with each other. [Non-local means](https://en.wikipedia.org/wiki/Non-local_means) and other similar methods are a very good fit for the KeOps engine, especially since @joanglaunes 's work on [higher-dimensional variables](https://www.kernel-operations.io/keops/_auto_benchmarks/plot_benchmark_high_dimension.html) that let us process larger patches relatively efficiently. 

As far as I can tell, the filter that you have in mind falls into the second category: Algorithm 1 in the [VideoWalk project page](https://ajabri.github.io/videowalk/) seems ideally suited to a KeOps implementation, especially if the resnet embedding of the patches is of dimension < 50. However, I'm not 100% sure that I fully understand your remark on temporal gaps.
Could you maybe write down a "**minimal working example**" in vanilla PyTorch that I could try to optimize with KeOps?

Documenting the use of KeOps outside of kernel methods / geometric deep learning is something that we'd really like to do in 2021, so that could be a neat example :-)
Best regards,
Jean

**Comment by vadimkantorov** [comment]
Traditionally for optical flow goals, correlation is computed between every two neighboring frames, so the context_length is 1, and temporal gap is always ~zero.

What videowalk computes is [`torch.einsum('ijklmn,ijkop->iklmnop', _k, _q)`](https://github.com/ajabri/videowalk/blob/master/code/utils/test_utils.py#L153)

For example (pseudocode):
```python
# 6840 is number of patches in a frame
# 256 is number of channels
# 69 is number of frames in video
# 21 is temporal context length
# 400 is number of each patch's nearest neigbors in space and time in the causal spatio-temporal neighborhood (i.e temporal context)

D = torch.zeros(6840, 6840) # spatial masking matrix
Q = torch.rand(256, 69, 6840)
K = torch.rand(256, 69, 21, 6840)
Att = f(Q, K, D) # [69, 21, 6840, 6840]
AfterTopkSoftmax = g(Att) # [69, 6840, 400]
```

**Comment by jeanfeydy** [comment]
I see!

To summarize, if I understand correctly:
1. Each frame is cut into 6840 patches. (I assume that this comes from e.g. a 114x60 grid.)
2. Each patch is represented by a CNN as a feature vector of size 256.
3. For every frame in the video (F=69) and every patch (P=6840), we want to compute an attention score with all the (P=6840) patches from the (T=21) preceding frames:
  **a.** The (F * P = 470k) query and (F * P * T = 10M) key vectors of size 256 are modified by small MLPs and turned into vectors of size S (= 256 too?).
  **b.** We compute a (F; P, S) @ (F; S, P * T) = (F; P, P * T) batched matrix-matrix product to compute attention scores: for every patch in the video, we compute "matching scores" with all patches from the (T=21) preceding frames. To reduce computation times and enforce a locality prior, we restrict the computations to a local neighborhood of each patch.
  **c.** We perform a  "TopK" reduction on the (F, P, P * T) attention matrix to retrieve, for each one of the (F * P = 470k) patches, the indices and scores of the K (= 400?) among (P * T = 144k)  closest matches. 

In this context, KeOps can come handy to **fuse steps 3.b and 3.c into one efficient layer**.
The main "selling point" of the library would be the **relatively easy implementation of an efficient locality pattern**. As detailed in [the doc](https://www.kernel-operations.io/keops/python/sparsity.html) and in [this tutorial](https://www.kernel-operations.io/keops/_auto_examples/pytorch/plot_grid_cluster_pytorch.html), we could use a pre-computed list of integers to handle this very efficiently. In practice, since the default CUDA block size of KeOps kernels is equal to 192, we could group patches in rectangular clusters of 16x12 patches, and specify a simple "1-to-9" block-wise sparsity mask, where every cluster interacts with the 3x3 closest clusters (+ along the time dimension, of course).

On the other hand, I must stress that the `.Kmin_argKmin(K=...)` reduction of KeOps is not optimized to work with values of K beyond 10-20. As detailed in [these benchmarks](https://www.kernel-operations.io/keops/_auto_benchmarks/plot_benchmark_KNN.html), this is a low-level issue: the scheme that is currently implemented in our engine stores the full list of K indices and values in the small CUDA registers. Large values cause register spilling. 
As a consequence, if you want to use K>50, the most efficient thing to do could be to fuse the estimation of the weights with the weighted mean itself, without any kind of K-NN truncation. (This is what the `SumSoftMaxWeight` [KeOps reduction](https://www.kernel-operations.io/keops/api/math-operations.html#reductions) is here for.) 
If the "variables to sum" are small enough (e.g. an [x, y, z] vector), this could be a viable option. 

What do you think?

**Comment by vadimkantorov** [comment]
3(a): In videowalk they even are not modified by small MLPs (no additional QK-projection), so dimensions of Q and K (and some underlying feature volume are all  256). A correction: K = 10 is used for topk. 400 is the number of neighbors pre-filtered by spatial locality pattern (e.g. values when I'm summing `D.sum(-1)`). The rest is correct.

Fusing it all together or even just getting the cost volume of shape `[69, 6840, 400]` would be fantastic. 

Then batching it all for processing independent videos is also a reasonable thing to think of (for consistency with the rest of PyTorch).

**Comment by jeanfeydy** [comment]
I see!
In this case, I believe that your computation will be a good fit for KeOps: `K=10` is perfect for our engine, as is the size of the variables that you are working with. The only complex thing to do here will be to define the correct integer "ranges" that define the block-wise sparsity mask. As discussed in #121 , point 5, I still haven't implemented a user-friendly "helper" function for 2D neighborhood structures, but I have detailed the relevant computation in #80. 
If you could provide me with a short working example (esp. with the "real" image sizes and patch extraction routines), this would be a good occasion to finally get this done :-)

**Comment by jeanfeydy** [comment]
Hi @vadimkantorov ,

Just realized that my last message was a bit ambiguous! Of course, I'll be happy to write the definition of the "ranges" for you, since it was on my to-do list anyway - it's a 1h-2h job. It's just that I would rather know precisely what are the dimensions of the frames / patch windows that you are using beforehand, so that I can try to optimize things well. If KeOps can deliver a meaningful speedup to your pipeline, that would be a very interesting indeed :-)

Best regards,
Jean

**Comment by vadimkantorov** [comment]
Sure! I will put up the shapes! Forgot yesterday

**Comment by vadimkantorov** [comment]
```
(Pdb) feats.shape
torch.Size([1, 256, 89, 60, 114])

# inputs to local attention + topk + softmax
(Pdb) additive_mask_out.shape
torch.Size([1, 1, 6840, 6840])

(Pdb) query.shape
torch.Size([1, 256, 69, 6840])

(Pdb) keys.shape
torch.Size([1, 256, 69, 21, 6840])


# outputs of local attention + topk + softmax
(Pdb) Ws.shape
torch.Size([69, 10, 6840])

(Pdb) Is.shape
torch.Size([69, 10, 6840])

```

**Comment by vadimkantorov** [comment]
Also, you might have been confused by the VideoWalk project page. Most likely it talks there about the training mode, where it only does batched matrix multiplication and it's fine. What I'm talking about here is what they currently do for inference and label propagation

**Comment by jeanfeydy** [comment]
Hi @vadimkantorov ,
Don’t worry, the information you sent me is perfect :-)
I have just updated the KeOps doc with some long overdue benchmarks and introduction pages: I will send you a first version of the “2D neighbourhoods” block wise sparsity masks in the next few days. It will fit right with @joanglaunes ’s non-local means tutorial too!
Best regards,
Jean

**Comment by vadimkantorov** [comment]
Also, further in the code it uses the computed weights/indices to aggregated the lablels stored in a different array (i.e. values in MHA terminology)
